npep/npep-126-egress-traffic-control.md

```diff
@@ -0,0 +1,90 @@
+# NPEP-126: Add northbound traffic support in (B)ANP API
+
+* Issue: [#126](https://github.com/kubernetes-sigs/network-policy-api/issues/126)
+* Status: Provisional
+
+## TLDR
+
+This NPEP proposes adding support for cluster egress (northbound) traffic control
+in the `AdminNetworkPolicy` and `BaselineAdminNetworkPolicy` API objects.
+
+## Goals
+
+* Implement egress traffic control towards external destinations (outside the cluster)
+  - NOTE: Currently there are no user stories for `Pass` action with northbound traffic
+    but users will get this for free since this action is anyways supported in the API.
+* Implement egress traffic control towards cluster nodes
+  - Currently the behaviour for policies defined around traffic from cluster
+    workloads (non-hostNetworked pods) towards nodes in the
+    cluster is undefined. See https://github.com/kubernetes-sigs/network-policy-api/issues/73.
+    - ANP currently supports only east-west traffic and this traffic flow cuts from
+    overlay to underlay which makes this part of the egress (northbound) use case.
+    - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+    - NOTE: Traffic towards nodes here includes traffic towards host-networked pods on that node
+      because a "node" resource encompasses all objects that share the host-networking resources
+* Implement egress traffic control towards k8s-apiservers
+  - An apiserver endpoint in this context is special in the sense that it can be any entity
+    including but not limited to a host-networked pod within the cluster OR external VMs or
+    bootstrap nodes running outside the cluster. This is why its a separate category goal.
```

Comment:
"bootstrap nodes" is an OpenShift concept that probably isn't meaningful to anyone else...

Reply:
ack changing this to external infra nodes instead

Reply:
done.